File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/model/dao/CurrencyInfoDao.java
package io.github.skywalkerdarren.simpleaccounting.model.dao;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

import io.github.skywalkerdarren.simpleaccounting.model.entity.CurrencyInfo;

@Dao
public interface CurrencyInfoDao {
    @Query("SELECT * FROM currency_info")
    List<CurrencyInfo> getInfos();

    @Query("SELECT * FROM currency_info WHERE name == :name")
    CurrencyInfo getInfo(String name);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void addInfo(CurrencyInfo info);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void updateInfo(CurrencyInfo info);

    @Query("DELETE FROM currency_info WHERE name == :name")
    void deleteInfo(String name);
}

File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/model/dao/CurrencyInfoDao.kt
package io.github.skywalkerdarren.simpleaccounting.model.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import io.github.skywalkerdarren.simpleaccounting.model.entity.CurrencyInfo

@Dao
interface CurrencyInfoDao {
    @get:Query("SELECT * FROM currency_info ORDER BY full_name_cn")
    val infos: LiveData<List<CurrencyInfo>>

    @Query("SELECT * FROM currency_info WHERE name == :name")
    fun getInfo(name: String): LiveData<CurrencyInfo>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addInfo(info: CurrencyInfo)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    suspend fun updateInfo(info: CurrencyInfo)

    @Query("DELETE FROM currency_info WHERE name == :name")
    suspend fun deleteInfo(name: String)

    @get:Query("SELECT * FROM currency_info WHERE name IN (SELECT name FROM currency WHERE favourite == 1)")
    val favInfos: LiveData<List<CurrencyInfo>>
}